test/app-tap/module_api.c

```diff
@@ -1711,6 +1715,83 @@ test_key_def_merge(struct lua_State *L)
 	key_def_check_merge(a_27, lengthof(a_27), b_27, lengthof(b_27),
 			    exp_27, lengthof(exp_27));
 
+	/*
+	 * Case 28: sort_order = 'asc' + sort_order = 'desc'.
+	 *
+	 * Interpretation: the same as for the case 3.
+	 */
+	box_key_part_def_t a_28[] = {
+		{{3, 0, "unsigned", NULL, NULL}},
+		{{1, 0, "unsigned", NULL, NULL}}, /* clash */
+	};
+	box_key_part_def_t b_28[] = {
+		{{1, 4, "unsigned", NULL, NULL}}, /* clash */
+		{{2, 0, "unsigned", NULL, NULL}},
+	};
+	box_key_part_def_t exp_28[] = {
+		{{3, 0, "unsigned", NULL, NULL}},
+		{{1, 0, "unsigned", NULL, NULL}}, /* coalesced */
+		{{2, 0, "unsigned", NULL, NULL}},
+	};
+	key_def_check_merge(a_28, lengthof(a_28), b_28, lengthof(b_28),
+			    exp_28, lengthof(exp_28));
+
+	/*
+	 * Case 29: sort_order = 'desc' + sort_order = 'asc'.
+	 *
+	 * Interpretation: the same as for the case 3.
+	 */
+	box_key_part_def_t a_29[] = {
+		{{3, 0, "unsigned", NULL, NULL}},
+		{{1, 4, "unsigned", NULL, NULL}}, /* clash */
+	};
+	box_key_part_def_t b_29[] = {
+		{{1, 0, "unsigned", NULL, NULL}}, /* clash */
+		{{2, 0, "unsigned", NULL, NULL}},
+	};
+	box_key_part_def_t exp_29[] = {
+		{{3, 0, "unsigned", NULL, NULL}},
+		{{1, 4, "unsigned", NULL, NULL}}, /* coalesced */
+		{{2, 0, "unsigned", NULL, NULL}},
+	};
+	key_def_check_merge(a_29, lengthof(a_29), b_29, lengthof(b_29),
+			    exp_29, lengthof(exp_29));
+
+	/*
+	 * Case 30: ascending with unicode collation and descending with
+	 * binary collation. In this case if we have different by binary
+	 * but the same by unicode strings, the same by unicode strings
+	 * will be grouped together and sorted in ascending order, but
+	 * the different by binary collation strings in these groups
+	 * will be sorted in descending order. Example:
+	 *
+	 * Tuple order by the unicode_ci collation if inserted in order:
+	 *  ['a'], ['A'], ['b'], ['B']
+	 *
+	 * Tuple order for the merged key_def with the binary collation:
+	 *  ['A'], ['a'], ['B'], ['b']
+	 *
+	 * Because 'a' has binary greater value (0x61), than 'A' (0x41).
+	 *
+	 * Interpretation: see the case 8.
+	 */
+	box_key_part_def_t a_30[] = {
+		{{3, 0, "unsigned", NULL,      NULL}},
+		{{1, 0, "unsigned", "unicode", NULL}}, /* clash */
+	};
+	box_key_part_def_t b_30[] = {
+		{{1, 4, "unsigned", "binary",  NULL}}, /* clash */
+		{{2, 0, "unsigned", NULL,      NULL}},
+	};
+	box_key_part_def_t exp_30[] = {
+		{{3, 0, "unsigned", NULL,      NULL}},
+		{{1, 0, "unsigned", "unicode", NULL}}, /* from <a> */
+		{{1, 4, "unsigned", "binary",  NULL}}, /* from <b> */
+		{{2, 0, "unsigned", NULL,      NULL}},
+	};
+	key_def_check_merge(a_30, lengthof(a_30), b_30, lengthof(b_30),
+			    exp_30, lengthof(exp_30));
```

Comment:
What is the idea behind this test case?

I mean, a test case is usually written with idea to verify some group of similar situations: they are either handled by the same code (if-branch) or they're similar logically.

It is not clear for me, what is the group that is assumed here.

I suppose that the idea is to verify that the logic that is described in the interpretation of the case 8 works for key parts with different sort order. Is it correct?

Or there is some logic that is specific for situations, when there are collations and the sort orders are different?

Feel free to ignore this comment. It is mostly for my understanding.

Reply:
Yes, the case is technically identical to the case 17, it just assures that the merged key part preserves its sort order.

Updated the case comment.